[WIP] Node Autoreplacement proposal

rzetelskik opened the pull request:
**Description of your changes:**
- Add enhancement proposal for Node Autoreplacement

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,114 @@
+## Proposal
+
+Create an InitContainer with a dedicated Scylla Operator command responsible for checking whether a given node is
+1. **not** first in the cluster
```
first in the sense of having index 0, or first bootstrapping node?

rzetelskik replied:
first bootstrapping node, see https://github.com/scylladb/scylla/blob/b503ec36c22856cf9e1d02ff0055767001be6f66/service/storage_service.cc#L240

tnozicka replied:
the answer was meant to be included in the document so it's not ambiguous

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,114 @@
+
+#### Story 2
+
+The user is running a ScyllaCluster on AWS/GCP cluster with a local storage. One of the instances gets restarted, which results in erasing the local storage.
```
I think restart keeps the local storage, suspend, hibernate and others don't

rzetelskik replied:
Explicit restart of an instance with local SSD is technically not allowed, but I've done the following:
1. Deploy a GKE cluster with 2 nodes
2. Wait for it to deploy
3. ssh to one of the instances and run `sudo shutdown -h now`
4. Wait for it to get back up again
And it resulted with losing storage and the pod logging `Startup failed: std::runtime_error (A node with address 10.55.254.64 already exists, cancelling join. Use replace_address if you want to replace this node.)`

tnozicka replied:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html
https://cloud.google.com/compute/docs/disks/local-ssd#data_persistence

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,114 @@
+
+## Design Details
+
+- The InitContainer uses the scylla-operator image.
```
I'd have expected of a description of who does it talk to to get the gossip information. If it's talking to the other nodes and not the local one, why should the code should live in a sidecar compared to the operator?

To which port does it talk to and how does it handle authentication?

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,114 @@
+
+Introducing an additional InitContainer, as well as a new shared Volume is required.
+
+## Alternatives
```
one alternative could be a detection from the local filesystem - either reading scylla files or creating a marker file when the node appears in bootstrap or becomes ready for the first time. also possibly talking to the local REST API.

rzetelskik commented:
@tnozicka I've changed the approach described in the proposal

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+Client is created with default config. Controller authenticates with the already known token. It communicates with all nodes of a given cluster. The conditions for creating the `ConfigMap` are met if the node's IP is among the returned hosts.
+
+`Sidecar` accesses the `ConfigMap` through a dedicated `Volume`.
+
```
details I am missing mentioned here:
- is the configmap optional or required
- is the CM live reloaded
- will any of it result in spec update / rollout
- possible races
- in which cases the controller talks to the nodes, what is the condition to not ask anymore
- how is the CM (holding the IP) updated when a node is replaced manually (with the replace label)
- what happens when a manual replace is already in progress
- what happens if a user config contains the replace-address-first-boot option (not that we should support it but the behaviour should be defined - like our takes precedence)

tnozicka replied:
I don't think I saw a mention of precedence with manual replace using the label

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+
+### Test Plan
+
+- e2e/integration tests
```
we don't have integration tests, just e2e

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+          - Wait for the `ScyllaCluster` to redeploy
+          - Verify `ScyllaCluster`
+    
+More fitting e2e tests could be created once the tests are run against a real distribution instead of Minikube.
```
How does it matter for this feature? I see the boundary as "a PV gets deleted", the node relation ship matters for the OrphanedPVController

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+          - Remove first node's PVC
+          - Remove first node's Pod
+          - Await PVC's and Pod's deletion
+          - Wait for the `ScyllaCluster` to redeploy
```
nit: redeploy seems slightly misleading here, `Wait for the ScyllaCluster to reconcile and become fully available again`

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+
+- Removing orphaned PVs. This is left for `OrphanedPVController`.
+
+## Proposal
```
explicitly mention what happens to a scyllacluster with 1 node (already bootstrapped)

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+     2.
+          - Create a `ScyllaCluster` with 2 nodes
+          - Wait for `ScyllaCluster` to deploy
+          - Turn on maintenance mode, Exec `systemctl stop scylla && rm -rf /var/lib/scylla` in first node's Pod
```
is this simulating hot swapping disks? would a scylla process be stopped in that case? otherwise this would look pretty similar to the first case where a PVC is deleted and it get new and empty disk.

rzetelskik replied:
You're right. I've removed the second test.

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+
+### Version Skew Strategy
+
+This enhancement only requires changes in the Scylla Operator image. There is no version skew involved.
```
this means operator N+1 vs. scyllacluster deployed with operator N and vice versa. Given you are changing the podSpec and creating a new object (CM) there is a skew in how the deployment looks like in each version. Describe what happens when the new operator sees the old deployment and vice versa.

tnozicka replied:
also what happens if the scyllacluster has lost the storage before the operator was upgraded

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+
+### Upgrade / Downgrade Strategy
+
+This enhancement adds a `ConfigMap` per `ScyllaCluster`'s Pod and extends `StatefulSetSpec` with a dedicated `Volume` to access the config.
```
This is meant in relation to an admin managing the scylla-operator. So something like: `No upgrade/downgrade action is needed, no manifests are changed.`

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+
+## Drawbacks
+
+- Controller needs to make API calls to Scylla nodes.
```
if there is a mitigation, like it's very infrequent, it's worth mentioning it as well

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+
+## Infrastructure Needed [optional]
+
+N/A
```
None.

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
````diff
@@ -0,0 +1,133 @@
+    replace-address-first-boot: <clusterIP>    
+```
+
+The config is later merged with generic config during Scylla setup, which effectively triggers running Scylla with `replace-address-first-boot` flag.
````
mention the config order

tnozicka commented:
I think the approach looks viable overall, needs a few details that should be flashed out in the proposal.

rzetelskik commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+### Notes/Constraints/Caveats [Optional]
+
+- It would be preferred to use [Immutable ConfigMaps](https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-immutable).
+- Having introduced this feature, we might consider removing the existing functionality of [replacing dead nodes manually](https://operator.docs.scylladb.com/master/nodeoperations/replace_node.html).
```
With this approach we would run Scylla with `replace-address-first-boot` set at all times after the node has bootstrapped, so I don't see any cases in which setting the label manually could change anything. Do you have anything specific in mind?

rzetelskik commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,133 @@
+
+#### Story 1
+
+The user is running a ScyllaCluster and looses a PVC bound to one of Scylla nodes. The node gets automatically replaced. 
```
I've removed that user story as it's almost a duplicate of the AWS/GCP one essentially

rzetelskik commented:
@tnozicka I've addressed all your comments. I'm not entirely sure about how this enhancement could be affected by running the manual replacement. I don't see any potential issues/races atm but I'm happy to discuss

tnozicka commented on enhancements/proposals/node-autoreplacement.md:
```diff
@@ -0,0 +1,128 @@
+The ConfigMap takes precedence over the generic config. In case `replace-address-first-boot` is already defined there, it is overwritten.
+
+The ConfigMap is optional and unavailable during the first start of a pod. It is not live reloaded and only read during the entrypoint setup, after every restart of a pod.
+The ScyllaCluster controller talks to the nodes whenever a generated ConfigMap differs from the existing one, which essentially comes down to two cases: after first boot and after a change to corresponding service's `ClusterIP`.
```
Won't the ConfigMaps differ through out the whole bootstrap process? That's currently like 1-2 minutes per node. Making manager API calls on every sync during that time is gonna be extremely costly. We need some condition that's based on the API object - e.g. could we only create the CM for nodes that are ready (and update always)? Guess we might not need to investigate the gossip at all then because a node ready implies presence in gossip.

rzetelskik commented:
Closing as stale
